### PR TITLE
[ISSUE #2576] Enhanced Configuration management

### PR DESCRIPTION
**Pull request by eight-nines**
Fixes #2576.

### Motivation

Enhancement the config manage for project

### Modifications

[1] Merge config feature branch into master branch

### Documentation

- Does this pull request introduce a new feature? no

### CODE REVIEW

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            configInfo.setPrefix(config.prefix());
+            configInfo.setMonitor(config.monitor());
+
+            return this.getConfig(configInfo);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void getConfig(Object object) throws IllegalAccessException, NoSuchFieldException, IOException {
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java:102]: suggest rename the method name,it is unclear or confusing with public <T> T getConfig(Class<?> clazz) method.  @eight-nines

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        ConfigFiled configFiled = field.getAnnotation(ConfigFiled.class);
+
+        Class<?> converter1 = configFiled.converter();
+        if (!converter1.equals(ConvertValue.DefaultConverter.class)) {
+            if (!classToConverter.containsKey(converter1)) {
+                try {
+                    ConvertValue<?> convertValue = (ConvertValue<?>) converter1.newInstance();
+                    register(convertValue, converter1);
+                } catch (Exception e) {
+                    e.printStackTrace();
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java:105]: It should use logger to replace e.printStackTrace().

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java (hunk 1 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java]: done, and comments, I'll add later

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java (hunk 2 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java]: done

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/converter/ObjectConverter.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    private void setValue() throws Exception {
+        boolean needReload = Boolean.FALSE;
+
+        for (Field field : this.clazz.getDeclaredFields()) {
+            if (Modifier.isStatic(field.getModifiers())) {
+                continue;
+            }
+            field.setAccessible(true);
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/converter/ObjectConverter.java:103]: Here should save the last accessible status and recovery it after custom operation.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+
+    /**
+     * Get the converter for the field
+     *
+     * @param field The field to be parsed
+     * @return the converter for the field
+     */
+    public static ConvertValue<?> getFieldConverter(Field field) {
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java:99]: What's difference between ConvertValue and FieldConverter?

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshGrpcBootstrap.java [hunk 5]
```diff
@@ -30,10 +30,12 @@ public class EventMeshGrpcBootstrap implements EventMeshBootstrap {
 
     private final Registry registry;
 
-    public EventMeshGrpcBootstrap(ConfigurationWrapper configurationWrapper, Registry registry) {
+    public EventMeshGrpcBootstrap(Registry registry) {
         this.registry = registry;
-        this.eventMeshGrpcConfiguration = new EventMeshGrpcConfiguration(configurationWrapper);
-        eventMeshGrpcConfiguration.init();
+
+        ConfigService configService = ConfigService.getInstance();
```
Comment (jonyangx) [inline eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshGrpcBootstrap.java:36]: @Test
    public void testEventMeshTcpBootstrap() {
        ConfigService configService = ConfigService.getInstance();
        EventMeshTCPConfiguration o = configService.buildConfigInstance(EventMeshTCPConfiguration.class);
        System.out.println(o);
    }

output:
EventMeshTCPConfiguration(eventMeshTcpServerPort=10000, eventMeshTcpIdleAllSeconds=60, eventMeshTcpIdleWriteSeconds=60, eventMeshTcpIdleReadSeconds=60, eventMeshTcpMsgReqnumPerSecond=15000, eventMeshTcpClientMaxNum=10000, eventMeshTcpGlobalScheduler=5, eventMeshTcpTaskHandleExecutorPoolSize=12, eventMeshTcpMsgDownStreamExecutorPoolSize=12, eventMeshTcpSessionExpiredInMills=60000, eventMeshTcpSessionUpstreamBufferSize=100, eventMeshTcpMsgAsyncRetryTimes=3, eventMeshTcpMsgSyncRetryTimes=1, eventMeshTcpMsgRetrySyncDelayInMills=500, eventMeshTcpMsgRetryAsyncDelayInMills=500, eventMeshTcpMsgRetryQueueSize=10000, eventMeshTcpRebalanceIntervalInMills=30000, eventMeshServerAdminPort=10106, eventMeshTcpSendBackEnabled=true, eventMeshTcpSendBackMaxTimes=3, eventMeshTcpPushFailIsolateTimeInMills=30000, gracefulShutdownSleepIntervalInMills=1000, sleepIntervalInRebalanceRedirectMills=200, eventMeshEventSize=1000, eventMeshEventBatchSize=10, gtc=EventMeshTCPConfiguration.TrafficShapingConfig(writeLimit=0, readLimit=10000, checkInterval=1000, maxTime=2000), ctc=EventMeshTCPConfiguration.TrafficShapingConfig(writeLimit=0, readLimit=2000, checkInterval=1000, maxTime=10000))
return all default value,but no load from properties file.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        if (Objects.isNull(configInfo.getPath()) || StringUtils.isEmpty(configInfo.getPath().trim())) {
+            object = FileLoad.getPropertiesFileLoad().getConfig(properties, configInfo);
+        } else {
+            String path = configInfo.getPath();
+            String filePath;
+            if (path.startsWith("classPath://")) {
+                filePath = Objects.requireNonNull(ConfigService.class.getResource("/" + path.substring(12))).getPath();
+            } else if (path.startsWith("file://")) {
+                filePath = path.substring(7);
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java:126]: I think length("file://")  to replace 7 is better.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigMonitorService.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            try {
+                Object object = ConfigService.getInstance().getConfig(configInfo);
+                if (configInfo.getObject().equals(object)) {
+                    continue;
+                }
+                configInfo.getObjectField().set(configInfo.getInstance(), object);
+                configInfo.setObject(object);
+                log.info("config connent : {}", object);
+            } catch (Exception e) {
+                log.error(e.getMessage(), e);
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigMonitorService.java:58]: log.error("load error", e) should better.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * Unified configuration Properties corresponding to eventmesh.properties
+     */
+    private Properties properties = new Properties();
+
+    @Getter
+    private String rootPath;
+
+    private final ConfigMonitorService configMonitorService = new ConfigMonitorService();
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java:45]: suggest add static .

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/Convert.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import org.apache.eventmesh.common.config.ConfigInfo;
+
+import java.util.Properties;
+
+/**
+ * Used to convert Config properties
+ */
+public class Convert {
+
+    public Object doConvert(ConfigInfo configInfo, Properties properties) {
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/Convert.java:29]: I think  it is better to use java generics .

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+            return;
+        }
+
+        for (Config config : configArray) {
+            populateConfig(object, clazz, config);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> T getConfig(ConfigInfo configInfo) throws IOException {
```
Comment (jonyangx) [inline eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java:115]: It should add unit test case.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/converter/ObjectConverter.java (hunk 3 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/converter/ObjectConverter.java]: Done! six six six!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java (hunk 4 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java:99]: ConvertValue is the top-level interface of the converter; FieldConverter is the converter of a specific field, for example, the converter of a field of type String is StringConverter

File: eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshGrpcBootstrap.java (hunk 5 above)
Comment (eight-nines) [reply eventmesh-runtime/src/main/java/org/apache/eventmesh/runtime/boot/EventMeshGrpcBootstrap.java:36]: In EventMeshStartup, we will set the path to the root configuration file and eventually resolve to EventMeshTCPConfiguration, UT can refer to EventMeshServerTest.
This design does not take into account the scenario of creating EventMeshTCPConfiguration directly in the main process, but for use in UT you can refer to EventMeshTCPConfigurationTest.
I will follow up with optimizations that have been documented.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java (hunk 6 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java]: Done!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        Object object;
+
+        if (Objects.isNull(configInfo.getPath()) || StringUtils.isEmpty(configInfo.getPath().trim())) {
+            object = FileLoad.getPropertiesFileLoad().getConfig(properties, configInfo);
+        } else {
+            String path = configInfo.getPath();
+            String filePath;
+            if (path.startsWith("classPath://")) {
```
[the remark this answers is not in the document]
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java]: Done!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigMonitorService.java (hunk 7 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigMonitorService.java]: Done!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java (hunk 8 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java]: Done!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java (hunk 10 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/ConfigService.java:118]: Done!

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/Convert.java (hunk 9 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/Convert.java:29]: I will follow up with optimizations that have been documented.

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java (hunk 4 above)
Comment (jonyangx) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java:99]: It's should have a goold name for ConvertValue. @eight-nines

**Review by jonyangx — APPROVED** [review]

File: eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java (hunk 4 above)
Comment (eight-nines) [reply eventmesh-common/src/main/java/org/apache/eventmesh/common/config/convert/ConverterMap.java:99]: Okay, I'll follow up!

**Pull request merged**